File: client/src/engine/tactical.ts
import { CoverType, MapId } from './types';

// Tactical terrain types (1 tile = 5 feet)
export type TacticalTerrain =
  | 'Floor'
  | 'Wall'
  | 'Dirt'
  | 'Stone'
  | 'Water'
  | 'Lava'
  | 'Grass'
  | 'Door'
  | 'StairsUp'
  | 'StairsDown';

export function isTacticalPassable(terrain: TacticalTerrain): boolean {
  return terrain !== 'Wall' && terrain !== 'Lava';
}

export function blocksSight(terrain: TacticalTerrain): boolean {
  return terrain === 'Wall' || terrain === 'Door';
}

// Tile specials
export type TileSpecial =
  | { type: 'Trap'; perceptionDc: number; damage: string }
  | { type: 'Chest'; looted: boolean }
  | { type: 'Lever'; activated: boolean }
  | { type: 'Sign'; text: string };

// A single tactical tile
export interface TacticalTile {
  terrain: TacticalTerrain;
  passable: boolean;
  cover: CoverType;
  special: TileSpecial | null;
  visible: boolean;
  explored: boolean;
}

export function createTile(terrain: TacticalTerrain): TacticalTile {
  return {
    terrain,
    passable: isTacticalPassable(terrain),
    cover: 'None',
    special: null,
    visible: false,
    explored: false,
  };
}

export function wallTile(): TacticalTile {
  return createTile('Wall');
}

export function floorTile(): TacticalTile {
  return createTile('Floor');
}

// Exit destinations
export type ExitDestination =
  | { type: 'WorldMap'; x: number; y: number }
  | { type: 'TacticalMap'; mapId: MapId; x: number; y: number };

// An exit point on a tactical map
export interface Exit {
  position: [number, number];
  destination: ExitDestination;
}

// A complete tactical map
export class TacticalMap {
  id: MapId;
  name: string;
  width: number;
  height: number;
  tiles: TacticalTile[][];
  elevation: number[][];
  spawnPoints: Array<[number, number]>;
  exits: Exit[];

  constructor(id: MapId, name: string, width: number, height: number) {
    this.id = id;
    this.name = name;
    this.width = width;
    this.height = height;
    this.tiles = [];
    this.elevation = [];
    for (let y = 0; y < height; y++) {
      const row: TacticalTile[] = [];
      const elevRow: number[] = [];
      for (let x = 0; x < width; x++) {
        row.push(floorTile());
        elevRow.push(0);
      }
      this.tiles.push(row);
      this.elevation.push(elevRow);
    }
    this.spawnPoints = [];
    this.exits = [];
  }

  getTile(x: number, y: number): TacticalTile | null {
    if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
      return this.tiles[y][x];
    }
    return null;
  }

  setTile(x: number, y: number, tile: TacticalTile): void {
    if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
      this.tiles[y][x] = tile;
    }
  }

  isPassable(x: number, y: number): boolean {
    const tile = this.getTile(x, y);
    return tile !== null && tile.passable;
  }

  exitAt(x: number, y: number): Exit | null {
    return this.exits.find((e) => e.position[0] === x && e.position[1] === y) ?? null;
  }
}

File: crates/core/src/tactical.rs
use serde::{Deserialize, Serialize};

use crate::types::*;

/// Terrain types for tactical maps (1 tile = 5 feet)
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum TacticalTerrain {
    Floor,
    Wall,
    Dirt,
    Stone,
    Water,
    Lava,
    Grass,
    Door,
    StairsUp,
    StairsDown,
}

impl TacticalTerrain {
    pub fn is_passable(&self) -> bool {
        !matches!(self, TacticalTerrain::Wall | TacticalTerrain::Lava)
    }

    pub fn blocks_sight(&self) -> bool {
        matches!(self, TacticalTerrain::Wall | TacticalTerrain::Door)
    }
}

/// Special features on a tile
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TileSpecial {
    Trap { perception_dc: u8, damage: String },
    Chest { looted: bool },
    Lever { activated: bool },
    Sign { text: String },
}

/// A single tile on a tactical map
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TacticalTile {
    pub terrain: TacticalTerrain,
    pub passable: bool,
    pub cover: CoverType,
    pub special: Option<TileSpecial>,
    pub visible: bool,
    pub explored: bool,
}

impl TacticalTile {
    pub fn new(terrain: TacticalTerrain) -> Self {
        let passable = terrain.is_passable();
        Self {
            terrain,
            passable,
            cover: CoverType::None,
            special: None,
            visible: false,
            explored: false,
        }
    }

    pub fn wall() -> Self {
        Self::new(TacticalTerrain::Wall)
    }

    pub fn floor() -> Self {
        Self::new(TacticalTerrain::Floor)
    }
}

/// Where an exit on a tactical map leads
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ExitDestination {
    WorldMap(u32, u32),
    TacticalMap(MapId, u16, u16),
}

/// An exit point on a tactical map
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Exit {
    pub position: (u16, u16),
    pub destination: ExitDestination,
}

/// A complete tactical map
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TacticalMap {
    pub id: MapId,
    pub name: String,
    pub width: u16,
    pub height: u16,
    pub tiles: Vec<Vec<TacticalTile>>,
    pub elevation: Vec<Vec<i8>>,
    pub spawn_points: Vec<(u16, u16)>,
    pub exits: Vec<Exit>,
}

impl TacticalMap {
    /// Create a new tactical map filled with floor tiles
    pub fn new(id: MapId, name: &str, width: u16, height: u16) -> Self {
        let tiles = vec![vec![TacticalTile::floor(); width as usize]; height as usize];
        let elevation = vec![vec![0i8; width as usize]; height as usize];
        Self {
            id,
            name: name.to_string(),
            width,
            height,
            tiles,
            elevation,
            spawn_points: Vec::new(),
            exits: Vec::new(),
        }
    }

    /// Get tile at position, None if out of bounds
    pub fn get_tile(&self, x: u16, y: u16) -> Option<&TacticalTile> {
        if x < self.width && y < self.height {
            Some(&self.tiles[y as usize][x as usize])
        } else {
            None
        }
    }

    /// Set tile at position
    pub fn set_tile(&mut self, x: u16, y: u16, tile: TacticalTile) {
        if x < self.width && y < self.height {
            self.tiles[y as usize][x as usize] = tile;
        }
    }

    /// Check if a position is passable
    pub fn is_passable(&self, x: u16, y: u16) -> bool {
        self.get_tile(x, y).map(|t| t.passable).unwrap_or(false)
    }

    /// Find an exit at the given position
    pub fn exit_at(&self, x: u16, y: u16) -> Option<&Exit> {
        self.exits.iter().find(|e| e.position == (x, y))
    }
}
